tests/test_infraction.py

```diff
@@ -0,0 +1,180 @@
+    return token
+
+
+@pytest.fixture()
+def coordinator(project, deployer, application, erc20, initiator, oz_dependency):
+    admin = deployer
+    contract = project.Coordinator.deploy(
+        application.address,
+        erc20.address,
+        FEE_RATE,
+        sender=deployer,
+    )
+
+    encoded_initializer_function = contract.initialize.encode_input(TIMEOUT, MAX_DKG_SIZE, admin)
+    proxy = oz_dependency.TransparentUpgradeableProxy.deploy(
+        contract.address,
+        deployer,
+        encoded_initializer_function,
+        sender=deployer,
+    )
+    proxy_contract = project.Coordinator.at(proxy.address)
+
+    proxy_contract.grantRole(contract.INITIATOR_ROLE(), initiator, sender=admin)
+    return proxy_contract
+
+
+@pytest.fixture()
+def global_allow_list(project, deployer, coordinator):
+    contract = project.GlobalAllowList.deploy(coordinator.address, sender=deployer)
+    return contract
+
+
+@pytest.fixture
+def infraction_collector(project, deployer, coordinator, application, oz_dependency):
+    contract = project.InfractionCollector.deploy(
+        coordinator.address, application.address, sender=deployer
+    )
+    proxy = oz_dependency.TransparentUpgradeableProxy.deploy(
+        contract.address,
+        deployer,
+        b"",
+        sender=deployer,
+    )
+    proxy_contract = project.InfractionCollector.at(proxy.address)
+    return proxy_contract
+
+
+def test_no_infractions(
+    erc20, nodes, initiator, global_allow_list, infraction_collector, coordinator
+):
+    for node in nodes:
+        public_key = gen_public_key()
+        coordinator.setProviderPublicKey(public_key, sender=node)
+        cost = coordinator.getRitualInitiationCost(nodes, DURATION)
+    erc20.approve(coordinator.address, cost, sender=initiator)
+    coordinator.initiateRitual(
+        nodes, initiator, DURATION, global_allow_list.address, sender=initiator
+    )
+    transcript = os.urandom(transcript_size(len(nodes), len(nodes)))
+    for node in nodes:
+        coordinator.postTranscript(0, transcript, sender=node)
+
+    with ape.reverts("Ritual must have failed"):
+        infraction_collector.reportMissingTranscript(0, nodes, sender=initiator)
+
+
+def test_report_infractions(
+    erc20, nodes, initiator, global_allow_list, infraction_collector, coordinator, chain
+):
+    for node in nodes:
+        public_key = gen_public_key()
+        coordinator.setProviderPublicKey(public_key, sender=node)
+        cost = coordinator.getRitualInitiationCost(nodes, DURATION)
+    erc20.approve(coordinator.address, cost, sender=initiator)
+    coordinator.initiateRitual(
+        nodes, initiator, DURATION, global_allow_list.address, sender=initiator
+    )
+    chain.pending_timestamp += TIMEOUT * 2
+    infraction_collector.reportMissingTranscript(0, nodes, sender=initiator)
+    for node in nodes:
+        assert infraction_collector.infractions(0, node) is True
+
```

Comment:
We should also test partial submissions i.e. some nodes submitted transcripts and some did not.

Reply:
^ One outstanding comment here. Something like half of the nodes submitted transcripts, but the other half did not.